cdap-watchdog/src/main/java/co/cask/cdap/logging/logbuffer/recover/LogBufferReader.java

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.logbuffer.recover;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.logging.logbuffer.LogBufferEvent;
+import co.cask.cdap.logging.logbuffer.LogBufferFileOffset;
+import co.cask.cdap.logging.serialize.LoggingEventSerializer;
+import com.google.common.io.Closeables;
+
+import java.io.BufferedInputStream;
+import java.io.Closeable;
+import java.io.DataInputStream;
+import java.io.EOFException;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+
+/**
+ * Reader to read log buffer files.
+ */
+public class LogBufferReader implements Closeable {
+  private static final String FILE_SUFFIX = ".buf";
+  private final int batchSize;
+  private final String baseDir;
+  private final long maxFileId;
+
+  private long currFileId;
+  private LogBufferEventReader eventReader;
+  private boolean skipFirstEvent;
+
+  /**
+   * Creates log buffer reader responsible for reading log buffer files.
+   *
+   * @param baseDir base directory for log buffer
+   * @param batchSize max number of log events to read in one batch
+   * @param currFileId current log buffer file id
+   * @param currPos position in current log buffer file
+   * @throws IOException if there is any error while opening file to read
+   */
+  public LogBufferReader(String baseDir, int batchSize, long currFileId, long currPos) throws IOException {
+    this.baseDir = baseDir;
+    this.batchSize = batchSize;
+    // if no checkpoints are written, currFileId and currPos will be -1. In that case the first event should not be
+    // skipped. However, if currFileId and currPos are non negative, that means first event should be skipped as
+    // atleast one log event has been persisted.
+    this.currFileId = currFileId < 0 ? 0 : currFileId;
+    this.skipFirstEvent = currFileId >= 0;
+    this.maxFileId = getMaxFileId(baseDir);
```

Comment:
Good to move the computation of max id to the startup to make the state being set clear